edalize/libero.py

```diff
@@ -0,0 +1,136 @@
+                        {'name': 'die',
+                         'type': 'String',
+                         'desc': 'FPGA device (e.g. MPF300TS)'},
+                        {'name': 'package',
+                         'type': 'String',
+                         'desc': 'FPGA package type (e.g. FCG1152)'},
+                        {'name': 'speed',
+                         'type': 'String',
+                         'desc': 'FPGA speed rating (e.g. -1)'},
+                        {'name': 'dievoltage',
+                         'type': 'String',
+                         'desc': 'FPGA die voltage (e.g. 1.0)'},
+                        {'name': 'range',
+                         'type': 'String',
+                         'desc': 'FPGA temperature range (e.g. IND)'},
+                        {'name': 'defiostd',
+                         'type': 'String',
+                         'desc': 'FPGA default IO std (e.g. "LVCMOS 1.8V"'},
+                        {'name': 'hdl',
+                         'type': 'String',
+                         'desc': 'Default HDL (e.g. "VERILOG"'}
+                    ]
+                    }
+
+    argtypes = ['vlogdefine', 'vlogparam']
+
+    tool_options_defaults = {
+        'speed': '-1',
+        'dievoltage': '1.0',
+        'range': 'IND',
+        'defiostd': 'LVCMOS 1.8V',
+        'hdl': 'VERILOG',
+    }
+
+    def _set_tool_options_defaults(self):
+        for key, default_value in self.tool_options_defaults.items():
+            if not key in self.tool_options:
+                logger.info("Set Libero tool option %s to default value %s"
+                            % (key, str(default_value)))
+                self.tool_options[key] = default_value
+
+    """ Initial setup of the class
+
+    This calls the parent constructor, but also identifies whether
+    the current system is using a Standard or Pro edition of Quartus.
+    """
+
+    def __init__(self, edam=None, work_root=None, eda_api=None):
+        if not edam:
+            edam = eda_api
+
+        super(Libero, self).__init__(edam, work_root)
+
+    """ Configuration is the first phase of the build
+
+    This writes the project TCL file. It first collects all
+    sources, IPs and constraints and then writes them to the TCL file along
+    with the build steps.
+    """
+
+    def configure_main(self):
+        self._set_tool_options_defaults()
+        (src_files, incdirs) = self._get_fileset_files(force_slash=True)
+        self.jinja_env.filters['src_file_filter'] = self.src_file_filter
+        self.jinja_env.filters['pdc_file_filter'] = self.pdc_file_filter
+
+        escaped_name = self.name.replace(".", "_")
+
+        template_vars = {
+            'name': escaped_name,
+            'src_files': src_files,
+            'incdirs': incdirs,
+            'tool_options': self.tool_options,
+            'toplevel': self.toplevel,
+            'generic': self.generic,
+            'prj_root': "../prj",
```

Comment:
All files needs to reside with in the work directory. If I get this correctly, files would end up in a sibling directory to the work root.

Reply:
Since the tcl file will be placed in generated dir (ex. `fusesoc_utils_blinky_1.0/polarfireeval_es-libero`, I've set the script to create a `prj` dir to save project in the `fusesoc_utils_blinky_1.0` dir (for example).
Is it ok?

Reply:
When FuseSoC sets up the directories, it will look like you describe. The problem is for users who use Edalize standalone since the prj directory will be outside of their workspace.

The best way would had been to generate the project files in `.`, like pretty much all the other EDA tools do but I noticed that Libero doesn't support that, so we have to put the `prj` directory below the work root, like

```
.
├── polarfireeval_es-libero
│   ├── prj                 <---- project dir will be created here      
│   ├── fusesoc_utils_blinky_1.0.eda.yml
│   ├── fusesoc_utils_blinky_1_0-project.tcl
│   └── fusesoc_utils_blinky_1_0-run.tcl
└── src
    └── fusesoc_utils_blinky_1.0
        ├── blinky.v
        └── polarfireeval
            ├── blinky_polarfireeval.v
            └── polarfireeval.pdc
```

Reply:
Done on last push.